common/src/main/java/io/netty/util/internal/telemetry/CycleBufferQueueEliminationStrategy.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util.internal.telemetry;
+
+import io.netty.util.internal.UnstableApi;
+
+import java.util.Arrays;
+import java.util.concurrent.atomic.AtomicInteger;
+
+@UnstableApi
+public class CycleBufferQueueEliminationStrategy implements QueueEliminationStrategy {
+    private static final long EMPTY_VALUE = -1L;
+
+    private final long[] latencies;
+    private final AtomicInteger pointer;
+
+    public CycleBufferQueueEliminationStrategy(int bufferSize) {
+        latencies = new long[bufferSize];
+        Arrays.fill(latencies, EMPTY_VALUE);
+        pointer = new AtomicInteger();
+    }
+
+    public long[] getLatencies() {
```

Comment:
Suggestion: Instead of returning an array which forces a copy to avoid unintended mutation, you can have a method like:

`public void forEachLatency(LatencyVisitor visitor)`

where `LatencyVisitor` is a functional interface like:

```java
public interface LatencyVisitor {
    /**
     * Callback for a latency value.
     * 
     * @param index Index of this value, starting from {@code 0}
     * @param size Total number o values that exist.
     * @param latencyNanos Latency value at {@code index}
     * @return {@code true} if next callback should be provided, {@code false} if iteration should stop early.
     */
    boolean latency(int index, int size, long latencyNanos); 
}
```

This is the pattern used for `ByteBuf#forEachByte()`